### PR TITLE
Attempting to fix slow NaiveBayes

### PR DESCRIPTION
**Pull request by jcalbert**
Three changes:

1) basic_extractor can accept a list of strings as well as a list of
('word','label') tuples.

2) BaseClassifier now has an instance variable _word_set which is a set
of tokens seen by the classifier.

1+2) BaseClassifier.extract_features passes _word_set to extractor
rather than the training set.

3)  NLTKClassifier.update adds new words to the _word_set.

### CODE REVIEW

**Comment by jcalbert** [comment]
As mentioned in a few issues, NaiveBayes is slow.

Test script, using the movie_reviews dataset:

```
import os, sys
from random import shuffle
pos_train = [(open('../movie_reviews/pos/'+fname,'r').read(),'pos') for fname in os.listdir('../movie_reviews/pos/')]
neg_train = [(open('../movie_reviews/neg/'+fname,'r').read(),'neg') for fname in os.listdir('../movie_reviews/neg/')]
all_train = pos_train + neg_train
shuffle(all_train)

#Old version
from textblob.classifiers_OLD import NaiveBayesClassifier as NBC

counts = [2**j for j in range(1,7)]
print 'Baseline: '
baseline_res = []
for n in counts:
    t = %timeit -o -q -n 1 -r 1 classifier = NBC(all_train[:n])
    s = str(n)
    print s + ' reviews' +  ' '*(5-len(s)) + str(round(t.best,3)) + ' sec'
    baseline_res.append(t.best)
    
from textblob.classifiers import NaiveBayesClassifier as NBC
print 'Modified: '
modified_res = []
for n in counts:
    t = %timeit -o -q -n 1 -r 1 classifier = NBC(all_train[:n])
    s = str(n)
    print s + ' reviews' +  ' '*(5-len(s)) + str(round(t.best,3)) + ' sec'
    modified_res.append(t.best)
    
```
with output:
```
Baseline: 
2 reviews    0.05 sec
4 reviews    0.137 sec
8 reviews    0.569 sec
16 reviews   2.57 sec
32 reviews   10.669 sec
64 reviews   42.275 sec

Modified: 
2 reviews    0.032 sec
4 reviews    0.056 sec
8 reviews    0.137 sec
16 reviews   0.338 sec
32 reviews   0.759 sec
64 reviews   1.803 sec
```

**Comment by jcalbert** [comment]
Relates to #63, #77 and #123.

**Comment by jcalbert** [comment]
From the travis-ci failures, it looks like the docstrings on `_get_words_from_dataset()` and `basic_extractor()` ask for more restricted training data that `test_classifiers.py` provides (many are not iterable).  Separate issue?

**Comment by sloria** [comment]
Thanks @jcalbert for the PR! I'll take a look at this over the weekend.

**Comment by jcalbert** [comment]
Chased down those old bugs.  TravisCI is complaining about a translation error, but I think that's Google's problem.

**Comment by komuher** [comment]
So guys?

**Comment by jcalbert** [comment]
As I mentioned above, the failed CI build is because of #161.  #162 fixes that.  Once it is merged then this (#136), #163, #167, and #170 should all pass CI after a rebase.

**Comment by sloria** [comment]
Thanks for the PR! And sorry for the delayed response.
